// overshom-wayforpay/dist/wfp/rest.api.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.restApi = void 0;
const axios_1 = require("axios");
const constants_1 = require("./constants");
const util_1 = require("./util");
const axios = axios_1.default.create({
    baseURL: constants_1.WAYFORPAY_API,
});
const sampleError = {
    code: 1,
    name: 'Unknown Error',
    reason: 'Unknown reason',
    messageForClient: 'Try again later. Please report problem if continue.',
    whoCanHelp: 'Product support',
};
const assertDetailedError = (err) => {
    if (typeof err !== 'object') {
        return false;
    }
    const hasAllKeys = Object.keys(sampleError).every(k => {
        return k in err;
    });
    return hasAllKeys;
};
class RestApi {
    baseCall(body, config) {
        return __awaiter(this, void 0, void 0, function* () {
            const { data } = yield axios.post('', body, config);
            try {
                util_1.validateSuccessResponse(data);
                return {
                    value: data,
                    error: undefined,
                };
            }
            catch (error) {
                if (!assertDetailedError(error)) {
                    throw new Error(JSON.stringify(error));
                }
                return {
                    value: undefined,
                    error,
                };
            }
        });
    }
}
exports.restApi = new RestApi();
