// overshom-wayforpay/dist/wfp/types/invoice-params.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvoiceParams = void 0;
const crypto = require("crypto");
const config_1 = require("../config");
class InvoiceParams {
    constructor({ amount, productName, productCount, productPrice, orderReference, merchantAccount, serviceUrl, merchantDomainName, MERCHANT_SECRET_KEY, paymentSystems, }) {
        this.transactionType = 'CREATE_INVOICE';
        this.currency = config_1.WFP_CONFIG.DEFAULT_PAYMENT_CURRENCY;
        this.apiVersion = 1;
        this.orderDate = new Date().getTime();
        this.holdTimeout = 60;
        this.merchantAccount = merchantAccount;
        this.serviceUrl = serviceUrl;
        this.merchantDomainName = merchantDomainName;
        const forHash = [
            this.merchantAccount,
            this.merchantDomainName,
            orderReference,
            this.orderDate,
            amount,
            this.currency,
            productName.join(';'),
            productCount.join(';'),
            productPrice.join(';'),
        ].join(';');
        this.merchantSignature = crypto
            .createHmac('md5', MERCHANT_SECRET_KEY)
            .update(forHash)
            .digest('hex');
        this.amount = amount;
        this.orderReference = orderReference;
        this.productName = productName;
        this.productCount = productCount;
        this.productPrice = productPrice;
        this.paymentSystems = paymentSystems === null || paymentSystems === void 0 ? void 0 : paymentSystems.join(';');
    }
}
exports.InvoiceParams = InvoiceParams;
